frontend/docsettings.lua

```diff
@@ -205,118 +209,101 @@ function DocSettings:open(docfile)
 end
 
 --- Serializes settings and writes them to `metadata.lua`.
-function DocSettings:flush()
-    -- write serialized version of the data table into one of
-    --  i) sidecar directory in the same directory of the document or
-    -- ii) history directory in root directory of KOReader
-    if not self.history_file and not self.sidecar_file then
-        return
-    end
+function DocSettings:flush(data)
+    -- Depending on the settings, doc_settings are saved to the book folder or
+    -- to koreader/docsettings folder. The latter is also a fallback for read-only book storage.
+    local serials = G_reader_settings:readSetting("document_metadata_folder", "doc") == "doc"
+        and { {self.doc_sidecar_dir, self.doc_sidecar_file},
+              {self.dir_sidecar_dir, self.dir_sidecar_file}, }
+         or { {self.dir_sidecar_dir, self.dir_sidecar_file}, }
 
-    -- If we can write to sidecar_file, we do not need to write to history_file anymore.
-    local serials = {}
-    if self.sidecar_file then
-        table.insert(serials, self.sidecar_file)
-    end
-    if self.history_file then
-        table.insert(serials, self.history_file)
-    end
-    self:ensureSidecar(self.sidecar)
-    local s_out = dump(self.data)
-    for _, f in ipairs(serials) do
+    local s_out = dump(data or self.data, nil, true)
+    for _, s in ipairs(serials) do
+        util.makePath(s[1])
```

Comment:
If i understand correctly, this would create a folder under `koreader/docsettings` for no reason even if using `doc`
Also `purgeDir` will only delete the immediate folder `koreader/docsettings/path/to/file.sdr/` only `file.sdr` is deleted, leaving empty folders.

Reply:
> If i understand correctly, this would create a folder under `koreader/docsettings` for no reason even if using `doc`

I believe no, we have a break of the loop if the "doc" sidecar has been saved.



> Also `purgeDir` will only delete the immediate folder `koreader/docsettings/path/to/file.sdr/` only `file.sdr` is deleted, leaving empty folders.

That's true.